File: leetcode/62-unique-paths.js
// https://leetcode.com/problems/unique-paths/ (top-interview-questions)

/** bottom-up approach (tabulation)
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function(m, n) {

  const tab = new Array(n).fill(1);
  
  for (let i = m - 2; i >= 0; i -= 1) {
    for (let j = n - 2; j >= 0; j -= 1) {
      tab[j] += tab[j+1];
    }
  }

  return tab[0];

};



/** top-down approach (memoization)
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths2 = function(m, n) {

  const memo = {};

  const dfs = function(r, c) {
    if (r === m-1 || c === n-1) {
      return 1;
    }

    const key = r + ',' + c;
    
    if (memo.hasOwnProperty(key)) {
      return memo[key];
    }

    const pathNums = dfs(r+1, c) + dfs(r, c+1);

    memo[key] = pathNums;

    return pathNums;

  }

  return dfs(0, 0);

};

File: leetcode/62-unique-paths.py
# https://leetcode.com/problems/unique-paths/ (top-interview-questions)

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        # top-down approach (memoization)

        cache = dict()

        def dfs(r, c):
            if r == m-1 or c == n-1:
                return 1
            if (r, c) in cache:
                return cache[(r, c)]

            unique_paths = dfs(r+1, c) + dfs(r, c+1)
            cache[(r, c)] = unique_paths
            return unique_paths

        return dfs(0, 0)
    

    def uniquePaths2(self, m: int, n: int) -> int:
        # bottom-up approach (tabulation)

        tab = [1] * n

        for _ in range(m-1):
            for i in reversed(range(n-1)):
                tab[i] += tab[i+1]

        return tab[0]

